map/common/cap.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2021  Masala
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package common
+
+import (
+	"github.com/katzenpost/katzenpost/core/crypto/eddsa"
+)
+
+var (
+	ReadCap  = []byte("read")
+	WriteCap = []byte("write")
+)
+
+// MessageID represents a storage address with Read/Write capability
+type MessageID [eddsa.PublicKeySize]byte
+
+// ReadPk returns the verifier of ReadCap for this ID
+func (m MessageID) ReadPk() *eddsa.PublicKey {
+	p := new(eddsa.PublicKey)
+	if err := p.FromBytes(m[:]); err != nil {
+		panic(err)
+	}
+	return p.Blind(ReadCap)
+}
+
+// WritePk returns the verifier of WriteCap for this ID
+func (m MessageID) WritePk() *eddsa.PublicKey {
+	p := new(eddsa.PublicKey)
+	if err := p.FromBytes(m[:]); err != nil {
+		panic(err)
+	}
+	return p.Blind(WriteCap)
+}
+
+func (m MessageID) Bytes() []byte {
+	return m[:]
+}
+
+type Cap interface {
+	Addr(addr []byte) MessageID
+}
+
+// ReadWriteCap describes a Capability that has Read and Write
+// capabilities and can return ReadOnly and WriteOnly capabilities
+type ReadWriteCap interface {
+	Cap
+	ReadOnlyCap
+	WriteOnlyCap
+	ReadOnly() ReadOnlyCap
+	WriteOnly() WriteOnlyCap
+}
+
+// ReadOnlyCap describes a Capability that has Read capability only.
+type ReadOnlyCap interface {
+	Cap
+	Read(addr []byte) *eddsa.BlindedPrivateKey
+}
+
+// ReadOnlyCap describes a Capability that has Write capability only.
+type WriteOnlyCap interface {
+	Cap
+	Write(addr []byte) *eddsa.BlindedPrivateKey
+}
+
+// PCap holds CapPk and Addr method
+type PCap struct {
```

Comment:
PCap is a name that doesn't tell the developer anything. Let's pick better names for things.

Reply:
PCap type provides method Addr and the capability public key for each type that inherits from it. Do you have a suggestion for a better name? PublicCapability? BaseCapability?